### PR TITLE
[FIZZ-152] Sync plus CreateDate/LastModifiedDate for extractor

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 1]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
     """
 
     courses_df = request_latest_courses_as_df(resource)
+    _sync_courses(courses_df, sync_db)
+    return courses_df
+
+
+def _sync_courses(courses_df: DataFrame, sync_db: sqlalchemy.engine.base.Engine):
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:128]: This is where the new sync work happens

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 2]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+    ----------
+    courses_df: DataFrame
+        a Courses API DataFrame with the current fetched data which
+        will be mutated, adding Hash and CreateDate/LastModifiedDate
+    sync_db: sqlalchemy.engine.base.Engine
+        an Engine instance for creating database connections
+    """
+    courses_df.set_index('id', inplace=True)
+
+    # compute hash from API call data
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:142]: create a hash for each row

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 3]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "GROUP BY id, Hash "
+            "HAVING COUNT(*) = 1"
+        )
+        con.execute("CREATE INDEX ID_INDEX ON Unmatched(id)")
+
+        # all rows start with CreateDate and LastModifiedDate initialized to "now",
+        #   but updated rows need the original CreateDate pulled from existing table
+        # Note: UPDATE-FROM is not available in sqlite until v3.33.0, thus the
+        #   double select goofiness
+        con.execute(
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:211]: Get the correct CreateDates from the regular table for updated records, and update the Unmatched table with them

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 4]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "    ) "
+            "    WHERE EXISTS ( "
+            "        SELECT * "
+            "        FROM Courses c "
+            "        WHERE c.id = Unmatched.id "
+            "    ) AND SyncNeeded = 1"
+        )
+
+        # fetch DataFrame with the changed record CreateDate updates
+        create_date_query = ("SELECT id, CreateDate "
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:226]: Pull those original CreateDates for just the updated records into a DF

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 5]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+                             "WHERE id IN ( "
+                             "    SELECT id FROM Unmatched "
+                             "    GROUP BY id "
+                             "    HAVING COUNT(*) > 1 "
+                             ") AND SyncNeeded = 1")
+        create_date_df = read_sql_query(create_date_query, con)
+        create_date_df.set_index('id', inplace=True)
+
+        # update the CreateDates of updated rows
+        courses_df.update(create_date_df)
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:237]: Update the CreateDates in the DF

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 6]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+
+        # now do deletes and inserts
+        con.execute(
+            "WITH "
+            #    changed rows CTE (from SyncNeeded side only)
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:242]: CTE to select only the change rows from Unmatched table

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 7]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "DELETE FROM Courses "
+            "WHERE id IN ( "
+            "    SELECT id from changedRows "
+            ")"
+        )
+
+        con.execute(
+            # insert new and changed data
+            "WITH "
+            #    changed rows CTE (from SyncNeeded side only)
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:263]: Right now duplicating this CTE to select only the change rows from Unmatched table. Need to either restructure or figure out how to share across executions with SQlAlchemy (didn't need to do this in script)

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 8]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "        SELECT * FROM Unmatched "
+            "        WHERE id IN ( "
+            "            SELECT id FROM Unmatched "
+            "            GROUP BY id "
+            "            HAVING COUNT(*) > 1 "
+            "        ) AND SyncNeeded = 1 "
+            "    )"
+
+
+            # delete the obsolete data
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:253]: Drop the changed rows from regular table

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 9]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            #    changed rows CTE (from SyncNeeded side only)
+            "    changedRows AS ( "
+            "        SELECT * FROM Unmatched "
+            "        WHERE id IN ( "
+            "            SELECT id FROM Unmatched "
+            "            GROUP BY id "
+            "            HAVING COUNT(*) > 1 "
+            "        ) AND SyncNeeded = 1 "
+            "    ),"
+            #    new rows CTE (also from SyncNeeded side)
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:272]: CTE to select only the new rows from Unmatched table

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 10]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "    newRows AS ( "
+            "        SELECT * FROM Unmatched "
+            "        WHERE id IN ( "
+            "            SELECT id FROM Unmatched "
+            "            GROUP BY id "
+            "            HAVING COUNT(*) = 1 AND SyncNeeded = 1 "
+            "        ) "
+            "    ) "
+
+            "INSERT INTO Courses "
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:282]: Insert new and changed rows from Unmatched into regular table

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 11]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "    SELECT * FROM Unmatched "
+            "    WHERE id IN ( "
+            "        SELECT id FROM changedRows "
+            "        UNION ALL "
+            "        SELECT id FROM newRows "
+            "    ) AND SyncNeeded = 1"
+        )
+
+        con.execute(
+            # reset SyncNeeded flag on main table
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:292]: Clear the SyncNeeded flag for newly inserted rows

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 12]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+            "UPDATE Courses "
+            "SET SyncNeeded = 0 "
+            "WHERE SyncNeeded != 0"
+        )
+
+        # finally, remove sync table
+        con.execute("DROP TABLE IF EXISTS Sync_Courses")
```
Comment (bradbanister) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:299]: cleanup temp tables

File: src/google-classroom-extractor/tests/api/courses/courses-initial.csv [hunk 13]
```diff
@@ -0,0 +1,9 @@
+id,name,section,descriptionHeading,description,room,ownerId,creationTime,updateTime,enrollmentCode,courseState,alternateLink,teacherGroupEmail,courseGroupEmail,guardiansEnabled,calendarId,SyncNeeded,CreateDate,LastModifiedDate
```
Comment (bradbanister) [inline src/google-classroom-extractor/tests/api/courses/courses-initial.csv:1]: initial test has regular table with one course with update, two courses that are in both but unchanged, and one course not in the sync table

File: src/google-classroom-extractor/tests/api/courses/courses-sync.csv [hunk 14]
```diff
@@ -0,0 +1,9 @@
+id,name,section,descriptionHeading,description,room,ownerId,creationTime,updateTime,enrollmentCode,courseState,alternateLink,teacherGroupEmail,courseGroupEmail,guardiansEnabled,calendarId
```
Comment (bradbanister) [inline src/google-classroom-extractor/tests/api/courses/courses-sync.csv:1]: initial test has sync table with one course with update, two courses that are in both but unchanged, and one course not in the regular table

File: src/google-classroom-extractor/tests/api/courses/sync.sql [hunk 15]
```diff
@@ -0,0 +1,120 @@
+-- Create Courses table
```
Comment (bradbanister) [inline src/google-classroom-extractor/tests/api/courses/sync.sql:1]: The script I was using for manual testing - will remove later

File: src/google-classroom-extractor/tests/api/courses/test_courses_sync.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (bradbanister) [inline src/google-classroom-extractor/tests/api/courses/test_courses_sync.py:1]: initial test - right now just shows that execution works without SQL exceptions

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py (hunk 2 above)
Comment (stephenfuqua) [reply src/google-classroom-extractor/google_classroom_extractor/api/courses.py:142]: You might be interested in in an alternative built right into Pandas... see [this notebook](https://github.com/Ed-Fi-Exchange-OSS/Ed-Fi-X-Fizz/blob/09c668c3435512688925638d10a45da5bd964b76/experimental/python/hashing.ipynb) I just uploaded.

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 17]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+    # push to temporary sync table
     courses_df.to_sql(
-        "Courses", sync_db, if_exists="append", index=False, chunksize=500
+        "Sync_Courses", sync_db, if_exists="replace", index=True, chunksize=1000
     )
-    # remove duplicates - leave only the most recent
+
     with sync_db.connect() as con:
+        # ensure main table exists
         con.execute(
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply src/google-classroom-extractor/google_classroom_extractor/api/courses.py:163]: bummer

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py [hunk 18]
```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
...
+        #   so SyncNeeded will show which row is from which table
+        con.execute(
+            "CREATE TABLE Unmatched AS "
+            "SELECT * FROM ( "
+            "    SELECT * FROM Courses "
+            "    UNION ALL "
+            "    SELECT * FROM Sync_Courses "
+            ") "
+            "GROUP BY id, Hash "
+            "HAVING COUNT(*) = 1"
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/google_classroom_extractor/api/courses.py:203]: Suggestion... easier to read and edit this way:
```suggestion
            """
            CREATE TABLE Unmatched AS
            SELECT * FROM (
                SELECT * FROM Courses
                UNION ALL
                SELECT * FROM Sync_Courses
            )
           GROUP BY id, Hash
           HAVING COUNT(*) = 1
           ''"
```

File: src/google-classroom-extractor/google_classroom_extractor/mapping/assignment_submissions.py [hunk 19]
```diff
@@ -44,8 +44,8 @@ def submissions_to_assignment_submissions_dfs(
         Status: The status of the submission in relation to the late acceptance policy
         SubmissionDateTime: The date and time of the assignment submission
         LMSUserIdentifier: A unique numeric identifier assigned to the user
-        CreateDate: Date this record was created
-        LastModifiedDate: Date this record was last updated
+        SourceCreateDate: Date this record was created in the LMS
+        SourceLastModifiedDate: Date this record was last updated in the LMS
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/google_classroom_extractor/mapping/assignment_submissions.py:48]: Don't we need all four columns in the output file?

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py (hunk 2 above)
Comment (bradbanister) [reply src/google-classroom-extractor/google_classroom_extractor/api/courses.py:142]: Ok, sold thanks to the speed.  As long as it's stable across application launches.  The Python hash() function is not, and hash_pandas_object() is poorly documented. I'd seen hash_pandas_object() but didn't know if it used hash() under the hood, and had seen that it has trouble with some data types, probably not relevant though.

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py (hunk 18 above)
Comment (bradbanister) [reply src/google-classroom-extractor/google_classroom_extractor/api/courses.py:203]: 👍

File: src/google-classroom-extractor/google_classroom_extractor/mapping/assignment_submissions.py (hunk 19 above)
Comment (bradbanister) [reply src/google-classroom-extractor/google_classroom_extractor/mapping/assignment_submissions.py:48]: Yes, that was an early step to rename to "SourceXYZDate".  Still need to document and include the other two.

File: src/google-classroom-extractor/google_classroom_extractor/api/courses.py (hunk 2 above)
Comment (bradbanister) [reply src/google-classroom-extractor/google_classroom_extractor/api/courses.py:181]: It turned out that hash_pandas_object() is not always stable after all.  However, in playing around with pandas row serialization I found that to_json() is way faster than to_csv().  Switching to that plus switching to a popular non-crypto hash library - xxhash - gives much improved performance.  Added perf test to notebook.

**Comment by bradbanister** [comment]
You will need to delete the old sqlite db (if exists) before running.  Or at least delete the Courses table.

**Review by stephenfuqua — APPROVED** [review]
Looks good, pending QA.

**Pull request merged**